zproject/backends.py

```diff
@@ -729,17 +689,40 @@ def user_data(self, access_token: str, *args: Any, **kwargs: Any) -> Dict[str, s
 class AzureADAuthBackend(SocialAuthMixin, AzureADOAuth2):
     auth_backend_name = "AzureAD"
 
+class GoogleAuthBackend(SocialAuthMixin, GoogleOAuth2):
+    auth_backend_name = "Google"
+    name = "google"
+
+    def get_verified_emails(self, *args: Any, **kwargs: Any) -> List[str]:
+        verified_emails = []    # type: List[str]
+        response = kwargs["response"]
+        if 'email' not in response:
+            # User info not loaded, load it now.
+            access_token = kwargs["response"]["access_token"]
+            try:
+                details = self.user_data(access_token)
+            except (HTTPError, ValueError, TypeError):  # nocoverage
+                verified_emails = []
+                return verified_emails
```

Comment:
@timabbott Sorry for late reply here, we need to call the `user_data` here only for [this](https://github.com/HarshitOnGitHub/zulip/blob/20309d2b8d59779c5dc3a2a538eefbac9e64c3ae/zerver/tests/test_auth_backends.py#L297) test case which calls the `authenticate()` method directly instead of following the normal flow for social auth. For normal flow it works as you expected without calling `user_data()` function. I think that test case is of little use and we can remove the test case as well as this piece of code.

Reply:
I'm reluctant to do that test, but could we just do that setup work inside `patched_authenticate` or something?

Reply:
@timabbott I am in favor of just removing that test since it is not testing something that we are currently or likely to use in the codebase. If we take the `patch_authenticate()` approach then we will be actually testing `patch_authenticate()` rather than `authenticate()` itself, right?

Reply:
Just to be clear, the AuthBackendTest as a whole has as its purpose testing that the complex and difficult-to-reason-about logic for which authentication backends actually run with a given authenticate() call is actually used.  (It's difficult to reason about because `authenticate() has an unusual mechanism where it calls only those backends' authenticate methods that match the signature of the authenticate() arguments used in that particular call).  In my view, it's critical that we test that for all social auth backends in that test class.

Reply:
Oh, could we call parts of the social auth pipeline outside `verify_backend`?  Or possibly by changing `good_kwargs`, e.g. by extending `token_data_dict`?